actions/rules/action_state_to_metastate.go

```diff
@@ -0,0 +1,459 @@
+		return nil, err
+	}
+	stateList, err := act.getValueListFromEnumField(wit, workitem.SystemState)
+	if err != nil {
+		return nil, err
+	}
+	metastateList, err := act.getValueListFromEnumField(wit, workitem.SystemMetaState)
+	if err != nil {
+		return nil, err
+	}
+	if len(stateList) != len(metastateList) {
+		return nil, errs.Errorf("inconsistent number of states and metatstates in the current work item type (must be equal): %d states != %d metastates", len(stateList), len(metastateList))
+	}
+	stateToMetastateMap := make(map[string]string)
+	for idx := range stateList {
+		thisState, ok := stateList[idx].(string)
+		if !ok {
+			return nil, errs.New("state value in value list is not of type string")
+		}
+		thisMetastate, ok := metastateList[idx].(string)
+		if !ok {
+			return nil, errs.New("metastate value in value list is not of type string")
+		}
+		// only the first state<->metatstate mapping needs to be taken as
+		// per definition. So we're only adding a new entry here if there
+		// is not an entry already existing.
+		if _, ok := stateToMetastateMap[thisState]; !ok {
+			stateToMetastateMap[thisState] = thisMetastate
+		}
+	}
+	return stateToMetastateMap, nil
+}
+
+// getMetastateToStateMap returns the mapping from metastate to state values read from the template.
+func (act ActionStateToMetaState) getMetastateToStateMap(workitemTypeID uuid.UUID) (map[string]string, error) {
+	stateToMetastate, err := act.getStateToMetastateMap(workitemTypeID)
+	if err != nil {
+		return nil, err
+	}
+	metastateToStateMap := make(map[string]string)
+	for state, metastate := range stateToMetastate {
+		// this is important: we only add a new entry if there
+		// is not an entry already existing. Therefore, we're only
+		// using the first mapping, satisfying the req for getting the
+		// first matching state for a metastate.
+		if _, ok := metastateToStateMap[metastate]; !ok {
+			metastateToStateMap[metastate] = state
+		}
+	}
+	return metastateToStateMap, nil
+}
+
+func (act ActionStateToMetaState) addOrUpdateChange(changes change.Set, attributeName string, oldValue interface{}, newValue interface{}) change.Set {
+	for _, change := range changes {
+		if change.AttributeName == attributeName {
+			change.NewValue = newValue
+			return changes
+		}
+	}
+	newChanges := append(changes, change.Change{
+		AttributeName: attributeName,
+		OldValue:      oldValue,
+		NewValue:      newValue,
+	})
+	return newChanges
+}
+
+func (act ActionStateToMetaState) fuseChanges(c1 change.Set, c2 change.Set) change.Set {
+	for _, change := range c2 {
+		c1 = act.addOrUpdateChange(c1, change.AttributeName, change.OldValue, change.NewValue)
+	}
+	return c1
+}
+
+// OnChange executes the action rule. It implements rules.Action. Returns the Work Item with eventual changes applied.
+func (act ActionStateToMetaState) OnChange(newContext change.Detector, contextChanges change.Set, configuration string, actionChanges *change.Set) (change.Detector, change.Set, error) {
```

Comment:
I do not like the fact that `actionChanges` is a `pointer`. It is passed to `onStateChange`, `onBoardColumnChange` by reference. If one of `onStateChange`/`onBoardColumnChange` fails, there is a possibility that the original object would be modified even when there was an error. 

@michaelkleinhenz Can we get rid of `pointers` from `OnChange`, `onBoardColumnChange` and `onStateChange` methods? It is possible that original object is not modified when the error occurs in any of the methods but it would be better if we didn't use `pointers` where they are not necessary. This would help in debugging the failures as well.

Reply:
Ok, that is actually a good reason. I don't think that can happen with the state rule, but in case of an error, the `actionChanges` could be left in an undefined state. I updated the signature. f83ff16. Note that this also required some changes to the other existing rules.

Reply:
@michaelkleinhenz You're awesome. Thank you for changing it :tada: